### PR TITLE
Fix intra-node IP message fragmentation due to mismatched MTU between devices

### PR DESCRIPTION
**Pull request by joestringer**
```release-note
Fix issue where MTU-sized packets received the outside world would be fragmented within the node and dropped by L4+ policy. Existing endpoints must be re-deployed to fix the issue. For more information, see https://cilium.link/err-low-mtu .
```

Previously, we would set the MTU of endpoints and Cilium devices to 1450 by default, even if the underlying fabric connectivity would support up to 1500B. This was done to attempt to force everything to use an MTU that would fit inside tunnel packets encapsulated and sent over the device with access to the outside world. 

However, one unforseen side effect of this is that when a 1500B IP packet is sent into the node (outward-facing device with MTU 1500) from the world, and it is routed through to the `cilium_host` device with MTU 1450, the node itself will fragment such messages into multiple IP packets before the traffic hits Cilium. Despite the node theoretically being able to handle such larger (1500B) packets, the packets are unnecessarily fragmented. As a result, when latter fragments hit the policy handling part of the datapath, we are unable to extract the port properly (as the packet does not contain it), and L4-L7 policies cannot be applied to the packets. As a result, they are dropped. For web services this looks like a hang.

This PR proposes to raise the MTU of all devices to 1500, while introducing a default route in each endpoint that lowers the MTU to 1450. The intention of this is to retain the lower MTU behaviour for packet transmission outside the endpoint, while allowing larger packets to be received.

~~The PR also adds a check in the IPv4 paths so that when policy is enabled, all fragments are dropped with an obvious reason (rather than performing a lookup with a nonsense port).~~
_Dropped for now as it's an orthogonal change._

Don't be too scared of the large LoC change, it's mostly an upgrade of the `github.com/vishvananda/netlink` library version which adds support for route MTU.

Todo:
* [x] Clean CI run
* [x] Test upgrade
* [x] Test multi-node
* [x] Document upgrade
  * [x] Set up link for `https://cilium.link/err-low-mtu`

Potential followups:
* Add some kind of MTU API for Cilium to allow other parts such as scripts or plugins to query (or set?) the MTU so that everything interacting with the Cilium daemon can consistently use the same MTU  (would help to support higher MTU cases)
* Look at raising the MTU of all devices and creating a high-MTU route for local destinations.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4323)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: pkg/mtu/mtu.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+func DetectMTU() (err error) {
+	mtu := MaxMTU
+	devname := ""
+
+	links, err := netlink.LinkList()
+	for _, link := range links {
+		attrs := link.Attrs()
+
+		skip := true
+		for _, prefix := range deviceMTUPrefixes {
```
Comment (tgraf) [inline pkg/mtu/mtu.go]: Moving this loop into a function will make the skip logic simpler and easier to understand

File: pkg/node/manager.go [hunk 2]
```diff
@@ -191,6 +192,12 @@ func replaceNodeRoute(ip *net.IPNet) {
 	}
 
 	route := netlink.Route{LinkIndex: link.Attrs().Index, Dst: ip, Gw: via, Src: local}
+	// If the route includes the local address, then the route is for
+	// local containers and we can use a high MTU for transmit. Otherwise,
+	// it needs to be able to fit within the MTU of tunnel devices.
+	if !ip.Contains(local) {
```
Comment (tgraf) [inline pkg/node/manager.go:198]: Nice, this is beautiful

**Review by tgraf — DISMISSED** [review]
This looks awesome!

**Comment by joestringer** [comment]
I tested this with Multi-node. As is, traffic that comes from an endpoint towards an endpoint on another node delivers up to 1500B packets into the cilium_vxlan device apparently even in the TCP case, and during encapsulation, these packets are split across multiple vxlan packets so it always ends up generating two packets, 1514B + 84B for each packet from the endpoint. Maybe this is a slightly better compromise than what we have today where packets are being dropped due to fragmentation within the node, but still not really where we want to get to.

File: pkg/node/manager.go (hunk 2 above)
Comment (joestringer) [reply pkg/node/manager.go:198]: If only it worked as intended ;-)

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Managed to simplify it by setting the MTU of the default routes inside CNI containers to the tunnel MTU (ie, 1450). Note that if you log into a pod created with these settings, you might not see the MTU setting in `ip route` output, depending on the version of the `ip` command that exists within the container. However, the setting is actually applied.

File: pkg/node/manager.go (hunk 2 above)
Comment (joestringer) [reply pkg/node/manager.go:198]: I retained this, as routes to non-local destinations should actually use the lower MTU. It doesn't affect endpoint MTU, however.

**Comment by joestringer** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
Links is set up and points to http://docs.cilium.io/en/v1.1/install/upgrade/ for now. We can adjust it freely.

**Pull request merged**